metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py

```diff
@@ -0,0 +1,911 @@
+    def get_workspaces(self):
+        groups = self.get_groups()
+        workspaces = [
+            PowerBiAPI.Workspace(
+                id=workspace.get("id"),
+                name=workspace.get("name"),
+                state="",
+                datasets={},
+                dashboards=[],
+            )
+            for workspace in groups.get("value", [])
+            if workspace.get("type", None) == "Workspace"
+        ]
+        return workspaces
+
+    # flake8: noqa: C901
+    def get_workspace(
+        self, workspace_id: str, reporter: PowerBiDashboardSourceReport
+    ) -> Workspace:
+        """
+        Return Workspace for the given workspace identifier i.e. workspace_id
+        """
+        scan_create_endpoint = PowerBiAPI.API_ENDPOINTS[Constant.SCAN_CREATE]
+        scan_create_endpoint = scan_create_endpoint.format(
+            POWERBI_ADMIN_BASE_URL=PowerBiAPI.ADMIN_BASE_URL
+        )
+
+        def create_scan_job():
+            """
+            Create scan job on PowerBi for the workspace
+            """
+            request_body = {"workspaces": [workspace_id]}
+
+            res = requests.post(
+                scan_create_endpoint,
+                data=request_body,
+                params={
+                    "datasetExpressions": True,
+                    "datasetSchema": True,
+                    "datasourceDetails": True,
+                    "getArtifactUsers": True,
+                    "lineage": True,
+                },
+                headers={Constant.Authorization: self.get_access_token()},
+            )
+
+            if res.status_code not in (200, 202):
+                message = f"API({scan_create_endpoint}) return error code {res.status_code} for workspace id({workspace_id})"
+
+                LOGGER.warning(message)
+
+                raise ConnectionError(message)
+            # Return Id of Scan created for the given workspace
+            id = res.json()["id"]
+            LOGGER.info("Scan id({})".format(id))
+            return id
+
+        def wait_for_scan_to_complete(scan_id: str, timeout: int) -> Any:
+            """
+            Poll the PowerBi service for workspace scan to complete
+            """
+            minimum_sleep = 3
+            if timeout < minimum_sleep:
+                LOGGER.info(
+                    f"Setting timeout to minimum_sleep time {minimum_sleep} seconds"
+                )
+                timeout = minimum_sleep
+
+            max_trial = timeout // minimum_sleep
+            LOGGER.info(f"Max trial {max_trial}")
+            scan_get_endpoint = PowerBiAPI.API_ENDPOINTS[Constant.SCAN_GET]
+            scan_get_endpoint = scan_get_endpoint.format(
+                POWERBI_ADMIN_BASE_URL=PowerBiAPI.ADMIN_BASE_URL, SCAN_ID=scan_id
+            )
+
+            LOGGER.info(f"Hitting URL={scan_get_endpoint}")
+
+            trail = 1
+            while True:
+                LOGGER.info(f"Trial = {trail}")
+                res = requests.get(
```

Comment:
any reason to wrap this in another method that can try : except and potentially issue a retry as well?

Reply:
_create_scan_job_ is creating a job and _wait_for_scan_to_complete_ is waiting for a job to be completed. so no need to wrap it in another try : except